File: app/src/main/java/com/fermimn/gamewishlist/fragments/WishlistFragment.java
package com.fermimn.gamewishlist.fragments;

import android.os.AsyncTask;
import android.os.Bundle;
import android.util.Pair;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentActivity;
import androidx.lifecycle.Observer;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.DividerItemDecoration;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.fermimn.gamewishlist.App;
import com.fermimn.gamewishlist.R;
import com.fermimn.gamewishlist.adapters.GamePreviewListAdapter;
import com.fermimn.gamewishlist.models.Game;
import com.fermimn.gamewishlist.models.GamePreview;
import com.fermimn.gamewishlist.models.GamePreviewList;
import com.fermimn.gamewishlist.viewmodels.WishlistViewModel;

@Deprecated
public class WishlistFragment extends Fragment {

    @SuppressWarnings("unused")
    private static final String TAG = WishlistFragment.class.getSimpleName();

    private RecyclerView mRecyclerView;
    private GamePreviewListAdapter mAdapter;
    private WishlistViewModel mViewModel;
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private GamePreviewList mWishlist;

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater,
                             @Nullable ViewGroup container,
                             @Nullable Bundle savedInstanceState) {

        View view = inflater.inflate(R.layout.fragment_wishlist, container, false);

        mWishlist = new GamePreviewList();

        FragmentActivity activity = getActivity();
        if (activity != null) {
            // attach ViewModel to the Fragment
            mViewModel = new ViewModelProvider(getActivity()).get(WishlistViewModel.class);

            // init wishlist
            GamePreviewList wishlist = mViewModel.getWishlist().getValue();
            if (wishlist != null) {
                mWishlist.addAll(wishlist);
            }
        }

        activity = getActivity();
        if (activity != null) {

            // update RecyclerView UI after changes
            mViewModel.getWishlist().observe(activity, new Observer<GamePreviewList>() {
                @Override
                public void onChanged(GamePreviewList wishlist) {
                    int prevElements = mWishlist.size();
                    mWishlist.clear();
                    mWishlist.addAll(wishlist);

                    mAdapter.notifyDataSetChanged();
                    if (mAdapter.getItemCount() - prevElements > 0) {
                        mRecyclerView.smoothScrollToPosition(mAdapter.getItemCount() - 1);
                    }
                }
            });

            // show Toast during RecyclerView state changes
            mViewModel.isUpdating().observe(activity, new Observer<Pair<GamePreview, Boolean>>() {
                @Override
                public void onChanged(Pair<GamePreview, Boolean> updatedGame) {
                    if (updatedGame.first != null) {
                        if (updatedGame.second) {
                            Toast.makeText(
                                    getActivity(),
                                    "Downloading: " + updatedGame.first.getTitle(),
                                    Toast.LENGTH_SHORT
                            ).show();
                        } else {
                            Toast.makeText(
                                    getActivity(),
                                    "Aggiunto: " + updatedGame.first.getTitle(),
                                    Toast.LENGTH_SHORT
                            ).show();
                        }
                    }
                }
            });

        }

        mSwipeRefreshLayout = view.findViewById(R.id.swipe_to_refresh);
        mSwipeRefreshLayout.setOnRefreshListener(new SwipeRefreshLayout.OnRefreshListener() {
            @Override
            public void onRefresh() {

                // TODO: there could be memory leaks
                new AsyncTask<GamePreviewList, Integer, Boolean>() {
                    @Override
                    protected Boolean doInBackground(GamePreviewList... gamePreviewLists) {
                        for (int i = 0; i < mWishlist.size(); ++i) {
                            Game prev = (Game) mWishlist.get(i);
                            mViewModel.updateGame( mWishlist.get(i).getId() );
                            Game current = (Game) mWishlist.get(i);

                            if (current != null) {

                                int priceChanges = 0;
                                StringBuilder text = new StringBuilder();

                                // the game has been released
                                if (current.getNewPrice() != null && prev.getPreorderPrice() != null) {
                                    text.append( getString(R.string.notif_game_released) );
                                }

                                // lower new price
                                if (current.getNewPrice() != null && prev.getNewPrice() != null) {
                                    if (current.getNewPrice() < prev.getNewPrice()) {
                                        text.append( getString(R.string.notif_lower_new_price) );
                                        priceChanges++;
                                    }
                                }

                                // lower used price
                                if (current.getUsedPrice() != null && prev.getUsedPrice() != null) {
                                    if (current.getUsedPrice() < prev.getUsedPrice()) {
                                        text.append( getString(R.string.notif_lower_used_price) );
                                        priceChanges++;
                                    }
                                }

                                // lower digital price
                                if (current.getDigitalPrice() != null && prev.getDigitalPrice() != null) {
                                    if (current.getDigitalPrice() < prev.getDigitalPrice()) {
                                        text.append( getString(R.string.notif_lower_digital_price) );
                                        priceChanges++;
                                    }
                                }

                                // lower preorder price
                                if (current.getPreorderPrice() != null && prev.getPreorderPrice() != null) {
                                    if (current.getPreorderPrice() < prev.getPreorderPrice()) {
                                        text.append( getString(R.string.notif_lower_preorder_price) );
                                        priceChanges++;
                                    }
                                }

                                if (text.length() != 0) {
                                    text.deleteCharAt(text.length() - 1);

                                    if (priceChanges == 1) {
                                        App.sendOnUpdatesChannel(getContext(), current, text.toString(), null);
                                    } else {
                                        App.sendOnUpdatesChannel(getContext(), current, getString(R.string.notif_lower_prices), text.toString());
                                    }
                                }
                            }
                        }
                        return false;
                    }

                    @Override
                    protected void onPostExecute(Boolean refreshing) {
                        mSwipeRefreshLayout.setRefreshing(refreshing);
                    }
                }.execute();
            }
        });

        mRecyclerView = view.findViewById(R.id.wishlist);

        // all elements in the recyclerview have the same size
        mRecyclerView.setHasFixedSize(true);

        // init RecyclerView layout manager
        LinearLayoutManager layoutManager = new LinearLayoutManager( getActivity() );
        mRecyclerView.setLayoutManager(layoutManager);

        // add divider to RecyclerView
        DividerItemDecoration dividerItemDecoration =
                new DividerItemDecoration(getActivity(), layoutManager.getOrientation());
        mRecyclerView.addItemDecoration(dividerItemDecoration);

        // add Adapter to RecyclerView
        mAdapter = new GamePreviewListAdapter(getActivity(), mWishlist);
        mAdapter.setHasStableIds(true);
        mRecyclerView.setAdapter(mAdapter);

//        ItemTouchHelper itemTouchHelper =
//                new ItemTouchHelper(new SwipeToDeleteCallback(getActivity(), mAdapter, wishListViewModel));
//        itemTouchHelper.attachToRecyclerView(mRecyclerView);

        return view;
    }

}

File: app/src/main/java/com/fermimn/gamewishlist/fragments/WishlistFragment.kt
package com.fermimn.gamewishlist.fragments

import android.os.AsyncTask
import android.os.Bundle
import android.util.Log
import android.util.Pair
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentActivity
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import androidx.recyclerview.widget.DiffUtil
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.fermimn.gamewishlist.R
import com.fermimn.gamewishlist.custom_views.GamePreviewAdapter
import com.fermimn.gamewishlist.custom_views.GamePreviewRecyclerView
import com.fermimn.gamewishlist.models.GamePreview
import com.fermimn.gamewishlist.models.GamePreviewDiffUtilCallback
import com.fermimn.gamewishlist.models.GamePreviews
import com.fermimn.gamewishlist.utils.isNetworkAvailable
import com.fermimn.gamewishlist.viewmodels.WishlistViewModel
import java.lang.Exception
import java.lang.ref.WeakReference

class WishlistFragment : Fragment() {

    companion object {
        private val TAG: String = WishlistFragment::class.java.simpleName
    }

    private lateinit var swipeRefreshLayout: SwipeRefreshLayout
    private lateinit var recyclerView: GamePreviewRecyclerView
    private lateinit var adapter: GamePreviewAdapter

    private lateinit var viewModel: WishlistViewModel
    private val wishlist: GamePreviews = GamePreviews()

    override fun onCreateView(
            inflater: LayoutInflater,
            container: ViewGroup?,
            savedInstanceState: Bundle?
    ): View? {

        val view = inflater.inflate(R.layout.fragment_wishlist, container, false)

        swipeRefreshLayout = view.findViewById(R.id.swipe_to_refresh)
        recyclerView = view.findViewById(R.id.wishlist)

        activity?.let {
            viewModel = ViewModelProvider(it).get(WishlistViewModel::class.java)
            val gameList = viewModel.wishlist.value
            gameList?.let {
                wishlist.addAll(gameList)
            }

            adapter = GamePreviewAdapter(it, wishlist)
            recyclerView.adapter = adapter

            // update wishlist when the user add/remove a game
            viewModel.wishlist.observe(it, Observer { newItems ->

                val numItems = wishlist.size

                val diffResult = DiffUtil.calculateDiff(GamePreviewDiffUtilCallback(wishlist, newItems))

                wishlist.clear()
                wishlist.addAll(newItems)

                diffResult.dispatchUpdatesTo(adapter)

                if (wishlist.size == numItems+1) {
                    recyclerView.smoothScrollToPosition(adapter.itemCount-1)
                }

                Log.d(TAG, "wishlist updated")
            })

            // show toast while updating
            viewModel.isUpdating.observe(it, Observer<Pair<GamePreview?, Boolean>> { updatedGame ->

                Log.d(TAG, "updating wishlist...")

                if (updatedGame.first != null) {
                    val title: String = updatedGame.first?.title ?: ""

                    if (updatedGame.second) {
                        // TODO: remove hard-coded text
                        Toast.makeText(it, "Downloading: $title", Toast.LENGTH_SHORT).show()
                    } else {
                        // TODO: remove hard-coded text
                        Toast.makeText(it, "Added: $title", Toast.LENGTH_SHORT).show()
                    }
                }
            })

            swipeRefreshLayout.setOnRefreshListener {
                if (isNetworkAvailable(it)) {
                    SwipeToRefresh(it, swipeRefreshLayout).execute()
                } else {
                    swipeRefreshLayout.isRefreshing = false
                    Toast.makeText(
                            it,
                            resources.getText(R.string.toast_internet_not_available),
                            Toast.LENGTH_SHORT
                    ).show()
                }
            }

        }

        return view
    }

    // TODO: stop UpdateWorker while running this
    private class SwipeToRefresh(
            activity: FragmentActivity,
            swipeRefreshLayout: SwipeRefreshLayout
    ) : AsyncTask<Void?, Void?, Void?>() {

        companion object {
            private val TAG: String = SwipeToRefresh::class.java.simpleName
        }

        private val swipeRefreshLayout: WeakReference<SwipeRefreshLayout>
                = WeakReference(swipeRefreshLayout)
        private val viewModel: WishlistViewModel
                = ViewModelProvider(activity).get(WishlistViewModel::class.java)
        private val gamePreviews: GamePreviews

        init {
            gamePreviews = viewModel.wishlist.value ?: GamePreviews()
        }

        override fun doInBackground(vararg params: Void?): Void? {
            try {
                for (gamePreview in gamePreviews) {
                    Log.d(TAG, "updating game with id [${gamePreview.id}] ...")
                    viewModel.updateGame(gamePreview.id)
                }
            } catch (ex: Exception) {
                Log.e(TAG, "exception", ex)
            } finally {
                return null
            }
        }

        override fun onPostExecute(result: Void?) {
            swipeRefreshLayout.get()?.isRefreshing = false
        }

    }

}

